providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -72,435 +40,441 @@ static OSSL_FUNC_keymgmt_import_types_fn mlkem_imexport_types;
 {
-    if (key == NULL)
-        return 0;
+    const MLKEM_KEY *key1 = vkey1;
+    const MLKEM_KEY *key2 = vkey2;
 
-    if (key->keytype != MLKEM_KEY_TYPE_768)
-        return 0;
-
-    if (key->encoded_pubkey != NULL
-        && !ossl_param_build_set_octet_string(tmpl, params,
-                                              OSSL_PKEY_PARAM_PUB_KEY,
-                                              key->encoded_pubkey,
-                                              OSSL_MLKEM768_PUBLIC_KEY_BYTES))
+    if (!ossl_prov_is_running())
         return 0;
 
-    if (include_private
-        && key->encoded_privkey != NULL
-        && !ossl_param_build_set_octet_string(tmpl, params,
-                                              OSSL_PKEY_PARAM_PRIV_KEY,
-                                              key->encoded_privkey,
-                                              OSSL_MLKEM768_PRIVATE_KEY_BYTES))
-        return 0;
+    /* All we have that can be compared is key material */
+    if (!(selection & OSSL_KEYMGMT_SELECT_KEYPAIR))
+        return 1;
 
-    return 1;
+    return ossl_mlkem_vcompare_pubkeys(key1->vinfo, key1->pubkey, key1->prvkey,
+                                       key2->vinfo, key2->pubkey, key2->prvkey);
 }
 
-static int mlkem_export(void *key, int selection, OSSL_CALLBACK *param_cb,
+static int mlkem_export(void *vkey, int selection, OSSL_CALLBACK *param_cb,
                         void *cbarg)
 {
-    MLKEM768_KEY *mkey = key;
-    OSSL_PARAM_BLD *tmpl;
+    MLKEM_KEY *key = vkey;
+    vinfo_t v;
+    OSSL_PARAM_BLD *tmpl = NULL;
     OSSL_PARAM *params = NULL;
+    uint8_t *pubenc = NULL;
+    uint8_t *prvenc = NULL;
     int ret = 0;
 
-    debug_print("MLKEMKM export %p\n", key);
     if (!ossl_prov_is_running() || key == NULL)
         return 0;
 
     if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
         return 0;
 
-    tmpl = OSSL_PARAM_BLD_new();
-    if (tmpl == NULL)
+    /* Fail when no key material has yet been provided */
+    if (key == NULL || !have_keys(key)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
         return 0;
+    }
+    v = key->vinfo;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
-        int include_private = ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0);
+    pubenc = OPENSSL_malloc(v->pubkey_bytes);
+    if (pubenc == NULL)
+        goto err;
 
-        if (!key_to_params(mkey, tmpl, NULL, include_private))
+    if (key->prvkey != NULL
+        && (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
+        /*
+         * Allocated on the secure heap if configured, this is detected in
+         * ossl_param_build_set_octet_string(), which will then also use the
+         * secure heap.
+         */
+        prvenc = OPENSSL_secure_zalloc(v->prvkey_bytes);
+        if (prvenc == NULL)
             goto err;
     }
 
+    tmpl = OSSL_PARAM_BLD_new();
+    if (tmpl == NULL)
+        goto err;
+
+    /* The public key is always there if either is */
```

Comment:
OK, that I understand by now -- but shouldn't "selection" drive the decision which key/parts are exported? What if "OSSL_KEYMGMT_SELECT_PUBLIC_KEY" is not set? Doesn't this code export it regardless?

Reply:
Thanks. Will not return the public key when not requested.